_posts/2019-01-14-code-of-the-week.md

```diff
@@ -0,0 +1,127 @@
+
+**The summary: Datatypes are useful and important. Always using them for each
+exposed parameter will help you catch bugs earlier, and the more detailed
+your datatype is the better it will be at catching bugs.**
+
+After a very long introduction, let us come to the actual contribution. In
+calender week three we have a contribution from [David
+Hollinger](https://github.com/dhollinger). He invested a huge amount of
+time over a period of months into our [rsyslog
+module](https://forge.puppet.com/puppet/rsyslog#puppet-rsyslog). There are
+multiple great PRs that we could talk about here. Let us have a look at
+[the latest](https://github.com/voxpupuli/puppet-rsyslog/pull/106).
+
+You can create custom datatypes that are very complex/very strict. They are
+based on multiple scalar types. One very detailed example is this
+(types/actions/outputs/omelasticsearch.pp):
+
+```puppet
+ type Rsyslog::Actions::Outputs::Omelasticsearch = Struct[{
+   server               => Optional[Variant[String[1], Array[String[1]]]],
+   serverport           => Optional[Stdlib::Port],
+   healthchecktimeout   => Optional[Integer],
+   searchindex          => Optional[String[1]],
+   dynsearchindex       => Optional[Enum['on', 'off']],
+   searchtype           => Optional[String[1]],
+   dynsearchtype        => Optional[Enum['on', 'off']],
+   pipelinename         => Optional[String[1]],
+   dynpipelinename      => Optional[Enum['on', 'off']],
+   usehttps             => Optional[Enum['on', 'off']],
+   timeout              => Optional[Pattern[/^([0-9]+)(ms|s|m)$/]],
+   template             => Optional[String[1]],
+   bulkmode             => Optional[Enum['on', 'off']],
+   maxbytes             => Optional[Pattern[/^([0-9]+)[kKmMgGtT]$/]],
+   parent               => Optional[String[1]],
+   dynparent            => Optional[Enum['on', 'off']],
+   uid                  => Optional[String[1]],
+   pwd                  => Optional[String[1]],
+   errorfile            => Optional[Stdlib::Absolutepath],
+   'tls.cacert'         => Optional[Stdlib::Absolutepath],
+   'tls.mycert'         => Optional[Stdlib::Absolutepath],
+   'tls.myprivkey'      => Optional[Stdlib::Absolutepath],
+   bulkid               => Optional[String[1]],
+   dynbulkid            => Optional[Enum['on', 'off']],
+   writeoperation       => Optional[Enum['index', 'create']],
+   retryfailures        => Optional[Enum['on', 'off']],
+   retryruleset         => Optional[String[1]],
+   'ratelimit.interval' => Optional[Integer],
+   'ratelimit.burst'    => Optional[Integer],
+ }]
+```
+
+The rsyslog configuration has so many options, many of them can be nested. In
+the beginning of datatypes in Puppet, we would have used `Array` for the above
```

Comment:
Despite the way it reads, `Struct[{}]` isn't an `Array[Hash]`. Struct is an "Abstract" data type for defining the specific keys and values (or data types for the values) within a Hash.

Better wording would probably be:

```
The rsyslog configuration has so many options, many of them can be nested. In
the beginning of datatypes in Puppet, we would have used `Hash` for the above
requirement.  At this level the type is already better than the old `validate_hash()` 
function and everybody would be happy.

Reply:
For reference: https://puppet.com/docs/puppet/5.3/lang_data_abstract.html#struct

Reply:
thanks, updated it with your part.